# Data Preprocessing

The goal of this lab is to introduce you to data preprocessing techniques in order to make your data suitable for applying a learning algorithm.

## 1. Handling Missing Values

A common (and very unfortunate) data property is the ocurrence of missing and erroneous values in multiple features in datasets. For this exercise we will be using a data set about abalone snails.
The data set is contained in the Zip file you downloaded from Moodle (abalone.csv).

To determine the age of a abalone snail you have to kill the snail and count the annual
rings. You are told to estimate the age of a snail on the basis of the following attributes:
1. type: male (0), female (1) and infant (2)
2. length in mm
3. width in mm
4. height in mm
5. total weight in grams
6. weight of the meat in grams
7. drained weight in grams
8. weight of the shell in grams
9. number of annual rings (number of rings +1, 5 yields age)

However, the data is incomplete. Missing values are marked with −1.

In [2]:
import pandas as pd
# load data 
df = pd.read_csv("abalone.csv") #Should this not work please use the csv that was part of the zip file.
df.columns=['type','length','width','height','total_weight','meat_weight','drained_weight','shell_weight','num_rings']
df.head()

type  length  width  height  total_weight  meat_weight  drained_weight  \
0     0   0.350  0.265   0.090        0.2255       0.0995          0.0485   
1     1   0.530  0.420   0.135        0.6770       0.2565          0.1415   
2     0   0.440  0.365   0.125        0.5160       0.2155          0.1140   
3     2  -1.000  0.255   0.080        0.2050       0.0895          0.0395   
4     2   0.425  0.300   0.095        0.3515       0.1410          0.0775   

   shell_weight  num_rings  
0         0.070         -1  
1         0.210          9  
2         0.155         10  
3         0.055          7  
4         0.120          8

### Exercise 1.1

Compute the mean of of each numeric column and the counts of each categorical column, excluding the missing values.

In [3]:
##################

data_nomissing = df[df[df.columns]!= -1]
# missing values become Nan

mean = data_nomissing.mean()
count = data_nomissing.count()

print(mean)
print(count)

data_nomissing
##################

type              0.953534
length            0.523692
width             0.407955
height            0.139610
total_weight      0.828843
meat_weight       0.359263
drained_weight    0.180249
shell_weight      0.238604
num_rings         9.921756
dtype: float64
type              4089
length            4052
width             4052
height            4052
total_weight      4070
meat_weight       4051
drained_weight    4067
shell_weight      4074
num_rings         4077
dtype: int64


type  length  width  height  total_weight  meat_weight  drained_weight  \
0      0.0   0.350  0.265   0.090        0.2255       0.0995          0.0485   
1      1.0   0.530  0.420   0.135        0.6770       0.2565          0.1415   
2      0.0   0.440  0.365   0.125        0.5160       0.2155          0.1140   
3      2.0     NaN  0.255   0.080        0.2050       0.0895          0.0395   
4      2.0   0.425  0.300   0.095        0.3515       0.1410          0.0775   
...    ...     ...    ...     ...           ...          ...             ...   
4171   1.0   0.565  0.450   0.165        0.8870       0.3700          0.2390   
4172   0.0   0.590  0.440   0.135        0.9660       0.4390          0.2145   
4173   0.0   0.600  0.475   0.205        1.1760       0.5255          0.2875   
4174   1.0   0.625  0.485   0.150           NaN       0.5310          0.2610   
4175   0.0   0.710  0.555   0.195        1.9485       0.9455          0.3765   

      shell_weight  num_rings  
0           0.0700        NaN  
1           0.2100        9.0  
2           0.1550       10.0  
3           0.0550        7.0  
4           0.1200        8.0  
...            ...        ...  
4171        0.2490       11.0  
4172        0.2605       10.0  
4173        0.3080        9.0  
4174        0.2960       10.0  
4175        0.4950       12.0  

[4176 rows x 9 columns]

### Exercise 1.2

Compute the median of each numeric column,  excluding the missing values.

In [4]:
##################
median = data_nomissing.median()
print(median)
##################

type              1.00000
length            0.54500
width             0.42500
height            0.14000
total_weight      0.80175
meat_weight       0.33600
drained_weight    0.17050
shell_weight      0.23350
num_rings         9.00000
dtype: float64


### Exercise 1.3

Handle the missing values in a way that you find suitable. Think about different ways. Discuss dis-/advantages of your approach. Argue your choices.


First I kind of need to look at my data, how many missing values do I have?

In [5]:
rows_with_minus_one = df.apply(lambda row: (-1 in row.values), axis=1).sum()
print("Number of rows with at least one -1 value:", rows_with_minus_one)

Number of rows with at least one -1 value: 905


So apparently this is 905 elements with a missing value out of 4167 examples. Seem too many to delete/ignore those instances.
Is there any data in particular that is missing more than the others? Is there any relationship between these missing values?

In [6]:
import numpy as np
minus_one_counts = (df == -1).sum()
print(minus_one_counts)

type               87
length            124
width             124
height            124
total_weight      106
meat_weight       125
drained_weight    109
shell_weight      102
num_rings          99
dtype: int64


In [7]:
# Count missing values in each column
missing_counts_rows = data_nomissing.isna().sum(axis=1)

# Filter rows with more than three missing values
rows_with_many_missing = missing_counts_rows[missing_counts_rows > 3]

print("Indices of rows with more than three missing values:")
print(rows_with_many_missing.index)

print("\nCount of missing values in each of these rows:")
print(rows_with_many_missing)   

Indices of rows with more than three missing values:
Index([], dtype='int64')

Count of missing values in each of these rows:
Series([], dtype: int64)


So no any one piece of data is particularly as missing as the others. (We could further resarch into missing data by checking if there is any relationship between missing data all together)

Checking again our data, we don't really see any pattern about why data could be missing (meaning, our data is very likely not systematically missing, like in a form where single people don't need to fill when they got married or leave the square blank). Snails also don't have the best privacy laws, so data aggregation cannot explain this missing information either.

1. type: male (0), female (1) and infant (2)
2. length in mm
3. width in mm
4. height in mm
5. total weight in grams
6. weight of the meat in grams
7. drained weight in grams
8. weight of the shell in grams
9. number of annual rings (number of rings +1, 5 yields age)

Similarly, because our study objects are nails, what would be the purpose of finding a relationship between missing length in mm and width in mm?
As in, any one missing information was more likely related to a mistake/omission by the scientist measuring the snails. 
We could be interested in seeing if any one source of our data often missed some metric, and then trying to understan why (was the scientist lazy, or was this simply not a metric that was registered when they were working?) But this would be far more relevant on data for a real experiment than for our study case.

But an alternative could be to calculate the mean and just insert that to the missing values.


In [8]:
##################
#INSERT CODE HERE#
# 
##################

### Exercise 1.4

Perform Z-score normalization on every column (except the type of course!)

In [9]:
##################(
standard_derivation = data_nomissing.std()
Z_score = (data_nomissing[df.columns[1:]] - mean) / standard_derivation

Z_score = Z_score[df.columns]
Z_score
##################

type    length     width    height  total_weight  meat_weight  \
0      NaN -1.443993 -1.439534 -1.189003     -1.232135    -1.166955   
1      NaN  0.052442  0.121288 -0.110489     -0.310090    -0.461650   
2      NaN -0.695776 -0.432552 -0.350159     -0.638881    -0.645838   
3      NaN       NaN -1.540232 -1.428673     -1.273999    -1.211879   
4      NaN -0.820479 -1.087090 -1.069168     -0.974820    -0.980521   
...    ...       ...       ...       ...           ...          ...   
4171   NaN  0.343415  0.423382  0.608520      0.118767     0.048236   
4172   NaN  0.551253  0.322684 -0.110489      0.280100     0.358211   
4173   NaN  0.634389  0.675127  1.567199      0.708957     0.746803   
4174   NaN  0.842227  0.775825  0.249015           NaN     0.771511   
4175   NaN  1.548877  1.480712  1.327529      2.286541     2.633606   

      drained_weight  shell_weight  num_rings  
0          -1.205076     -1.211842        NaN  
1          -0.354425     -0.205594  -0.285597  
2          -0.605962     -0.600906   0.024243  
3          -1.287397     -1.319655  -0.905276  
4          -0.939820     -0.852468  -0.595436  
...              ...           ...        ...  
4171        0.537387      0.074718   0.334083  
4172        0.313290      0.157374   0.024243  
4173        0.981006      0.498780  -0.285597  
4174        0.738616      0.412530   0.024243  
4175        1.795070      1.842840   0.643923  

[4176 rows x 9 columns]

## 2. Preprocessing text (Optional)

One possible way to transform text documents into vectors of numeric attributes is to use the TF-IDF representation. We will experiment with this representation using the 20 Newsgroup data set. The data set contains postings on 20 different topics. The classification problem is to decide which of the topics a posting falls into. Here, we will only consider postings about medicine and space.

In [10]:
from sklearn.datasets import fetch_20newsgroups


categories = ['sci.med', 'sci.space']
raw_data = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42)
print(f'The index of each category is: {[(i,target) for i,target in enumerate(raw_data.target_names)]}')

The index of each category is: [(0, 'sci.med'), (1, 'sci.space')]


Check out some of the postings, might find some funny ones!

In [11]:
import numpy as np
idx = np.random.randint(0, len(raw_data.data))
print (f'This is a {raw_data.target_names[raw_data.target[idx]]} email.\n')
print (f'There are {len(raw_data.data)} emails.\n')
print(raw_data.data[idx])

This is a sci.space email.

There are 1187 emails.

From: aws@iti.org (Allen W. Sherzer)
Subject: Re: Why not give $1 billion to first year-long moon residents?
Organization: Evil Geniuses for a Better Tomorrow
Lines: 34

In article <C5sJDp.F23@zoo.toronto.edu> henry@zoo.toronto.edu (Henry Spencer) writes:

>>This prize isn't big enough to warrent developing a SSTO, but it is
>>enough to do it if the vehicle exists.

>Actually, there are people who will tell you that it *would* be enough
>to do SSTO development, if done privately as a cut-rate operation.  Of
>course, they may be over-optimistic.

In spite of my great respect for the people you speak of, I think their
cost estimates are a bit over-optimistic. If nothing else, a working SSTO
is at least as complex as a large airliner and has a smaller experience
base. It therefore seems that SSTO development should cost at least as
much as a typical airliner development. That puts it in the $3G to $5G
range.

>You can also assume that a 

Lets pick the first 10 postings from each category

In [12]:
idxs_med = np.flatnonzero(raw_data.target == 0)
idxs_space = np.flatnonzero(raw_data.target == 1)
idxs = np.concatenate([idxs_med[:10],idxs_space[:10]])
data = np.array(raw_data.data)
data = data[idxs]

<a href="http://www.nltk.org/">NLTK</a> is a toolkit for natural language processing. Take some time to install it and go through this <a href="http://www.slideshare.net/japerk/nltk-in-20-minutes">short tutorial/presentation</a>. (or use e.g. Google colab where the package is prepared already)

The downloaded package below is a tokenizer that divides a text into a list of sentences, by using an unsupervised algorithm to build a model for abbreviation words, collocations, and words that start sentences.

In [16]:
import nltk
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\juand\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt_tab.zip.


True

In [18]:
import nltk
import itertools

# Tokenize the sentences into words
tokenized_sentences = [nltk.word_tokenize(sent) for sent in data]
vocabulary_size = 1000
unknown_token = 'unknown'

In [19]:
# Count the word frequencies
word_freq = nltk.FreqDist(itertools.chain(*tokenized_sentences))
print (f"Found {len(word_freq.items())} unique words tokens.")

Found 1632 unique words tokens.


In [20]:
# Get the most common words and build index_to_word and word_to_index vectors
vocab = word_freq.most_common(vocabulary_size-1)
index_to_word = [x[0] for x in vocab]
index_to_word.append(unknown_token)
word_to_index = dict([(w,i) for i,w in enumerate(index_to_word)])
 
print (f"Using vocabulary size {vocabulary_size}." )
print (f"The least frequent word in our vocabulary is '{vocab[-1][0]}' and appeared {vocab[-1][1]} times.")

Using vocabulary size 1000.
The least frequent word in our vocabulary is 'NEEDS' and appeared 1 times.


### Exercise 2.1

Code your own TF-IDF representation function and use it on this dataset. (Don't use code from libraries. Build your own function with Numpy/Pandas). Use the formular TFIDF = TF * (IDF+1). The effect of adding “1” to the idf in the equation above is that terms with zero idf, i.e., terms that occur in all documents in a training set, will not be entirely ignored. The term frequency is the raw count of a term in a document. The inverse document frequency is the natural logarithm of the inverse fraction of the documents that contain the word.

In [21]:
from sklearn.feature_extraction.text import CountVectorizer
countvec = CountVectorizer()
df = pd.DataFrame(countvec.fit_transform(data).toarray(), columns=countvec.get_feature_names_out())

def tfidf(df):
    
    ##################
    #INSERT CODE HERE#
    ##################
    return None
    
    
rep = tfidf(df)

# Check if your implementation is correct
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(norm=None, smooth_idf=False, use_idf=True)
X_train = pd.DataFrame(vectorizer.fit_transform(data).toarray(), columns=countvec.get_feature_names_out())
answer=['No','Yes']
epsilon = 0.0001
if rep is None: 
  print (f'Is this implementation correct?\nAnswer: {answer[0]}')
if rep is not None:
  print (f'Is this implementation correct?\nAnswer: {answer[1*np.all((X_train - rep) < epsilon)]}')

Is this implementation correct?
Answer: No


In [ ]:
# an example of what to do with these similarities:


# analysis with tf-idf
from sklearn.metrics.pairwise import cosine_similarity

similiarities = cosine_similarity(rep, rep) # measure of the similarity of the direction of two vectors

In [ ]:
np.fill_diagonal(similiarities, 0)
max_ind = np.unravel_index(similiarities.argmax(), similiarities.shape)
similiarities[max_ind] # highest similarity of two documents